internal/controller/prefixclaim_controller.go

```diff
@@ -75,7 +75,49 @@ func (r *PrefixClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, nil
 	}
 
-	/* 1. check if the matching Prefix object exists */
+	/* 1. compute and assign the parent prefix if required */
+	// The current design will use prefixClaim.Status.ParentPrefix for storing the computed parent prefix, and as the source of truth for future parent prefix references
+	if prefixClaim.Status.ParentPrefix == "" /* parent prefix not yet computed/assigned */ {
+		if prefixClaim.Spec.ParentPrefix != "" {
+			// ParentPrefix takes precedence over ParentPrefixSelector
+			prefixClaim.Status.ParentPrefix = prefixClaim.Spec.ParentPrefix
+		} else if len(prefixClaim.Spec.ParentPrefixSelector) > 0 {
+			// The main idea is that we select one of the available parent prefixes as the ParentPrefix for all subsequent computation
+			//
+			// The existing algorithm for prefix allocation within a ParentPrefix remains unchanged
+
+			// fetch available prefixes from netbox
+			parentPrefixCandidates, err := r.NetboxClient.GetAvailablePrefixByParentPrefixSelector(prefixClaim.Spec.ParentPrefixSelector, prefixClaim.Spec.Tenant, prefixClaim.Spec.PrefixLength)
+			if err != nil || len(parentPrefixCandidates) == 0 {
+				errorMsg := fmt.Sprintf("no parent prefix can be obtained with the query conditions set in ParentPrefixSelector, err = %v, number of candidates = %v", err, len(parentPrefixCandidates))
+				if err := r.SetConditionAndCreateEvent(ctx, prefixClaim, netboxv1.ConditionParentPrefixComputedFalse, corev1.EventTypeWarning, errorMsg); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// we requeue as this might be a temporary exhausation
+				return ctrl.Result{Requeue: true}, nil
+			}
+
+			// set prefixClaim.Spec.ParentPrefix
+			// TODO(henrybear327): use best-fit algorithm to pick a parent prefix
+			parentPrefixCandidate := parentPrefixCandidates[0]
+			prefixClaim.Status.ParentPrefix = parentPrefixCandidate.Prefix
+		} else {
+			// this case should not be triggered anymore, as we have validation rules put in place on the CR
```

Comment:
Same as the above, I would like to log this in case we change more code in the future and this condition gets tripped somehow. 

What do you think? Or do you think if it's a given, we don't check for it :)

Reply:
we could add a test which checks that the prefix CR is not created if the ParentPrefix is not found. This way we could make sure the behavior doesn't occur until a case comes up where we need it.